app/src/main/java/pl/futurecollars/invoicing/db/file/FileBasedInvoiceRepository.java

```diff
@@ -0,0 +1,140 @@
+        this.jsonService = jsonService;
+    }
+
+    @Override
+    public Invoice save(Invoice invoice) {
+        UUID id = UUID.randomUUID();
+        try {
+            if (!this.getById(id).isPresent()) {
+                invoice.setId(id);
+                String json = jsonService.toJson(invoice);
+                fileService.writeToFile(json);
+                return invoice;
+            } else {
+                return save(invoice);
+            }
+        } catch (IOException e) {
+            throw new FileSystemException("File system processing error.");
+        }
+    }
+
+    @Override
+    public Optional<Invoice> getById(UUID id) {
+        List<Invoice> invoices;
+        try {
+            invoices = fileService.readFile()
+                .map(item -> {
+                    try {
+                        return jsonService.toObject(item, Invoice.class);
+                    } catch (JsonProcessingException e) {
+                        throw new FileSystemException("Json processing error.");
+                    }
+                }).collect(Collectors.toList());
+        } catch (IOException e) {
+            throw new FileSystemException("File system processing error.");
+        }
+
+        for (Invoice invoice : invoices) {
+            if (invoice.getId().equals(id)) {
+                return Optional.ofNullable(invoice);
+            }
+        }
+        return Optional.ofNullable(null);
+    }
+
+    @Override
+    public List<Invoice> getAll() {
+        try {
+            return fileService.readFile()
+                .map(item -> {
+                    try {
+                        return jsonService.toObject(item, Invoice.class);
+                    } catch (JsonProcessingException e) {
+                        throw new FileSystemException("Json processing error.");
+                    }
+                }).collect(Collectors.toList());
+        } catch (IOException e) {
+            throw new FileSystemException("File system processing error.");
+        }
+    }
+
+    @Override
+    public Invoice update(UUID id, Invoice updatedInvoice) {
+        List<Invoice> invoices;
+        try {
+            invoices = fileService.readFile()
+                .map(item -> {
+                    try {
+                        return jsonService.toObject(item, Invoice.class);
+                    } catch (JsonProcessingException e) {
+                        throw new FileSystemException("Json processing error.");
+                    }
+                }).collect(Collectors.toList());
+        } catch (IOException e) {
+            throw new FileSystemException("File system processing error.");
+        }
+
+        for (Invoice invoice : invoices) {
+            if (invoice.getId().equals(id)) {
+                int index = invoices.indexOf(invoice);
+                invoices.set(index, updatedInvoice);
+                return updatedInvoice;
+            }
+        }
+        return null;
```

Comment:
jaki jest scenariusz, że zwrócimy null? Jeżeli nie ma danego invoica to już lepiej rzucić wyjatkiem